src/databricks/labs/ucx/hive_metastore/locations.py

```diff
@@ -125,64 +164,59 @@ def __init__(
     @cached_property
     def _mounts_snapshot(self) -> list['Mount']:
         """Returns all mounts, sorted by longest prefixes first."""
-        return sorted(self._mounts_crawler.snapshot(), key=lambda _: len(_.name), reverse=True)
+        return sorted(self._mounts_crawler.snapshot(), key=lambda _: (len(_.name), _.name), reverse=True)
 
     def _external_locations(self) -> Iterable[ExternalLocation]:
-        min_slash = 2
-        external_locations: list[ExternalLocation] = []
+        trie = LocationTrie()
         for table in self._tables_crawler.snapshot():
-            location = table.location
-            if not location:
+            table = self._resolve_location(table)
+            if not table.location:
                 continue
-            # TODO: refactor this with LocationTrie
-            if location.startswith("dbfs:/mnt"):
-                location = self.resolve_mount(location)
-            if not location:
+            trie.insert(table)
+        queue = list(trie.children.values())
+        external_locations = []
+        while queue:
+            curr = queue.pop()
+            num_children = len(curr.children)  # 0 - take parent
+            if curr.location and (num_children > 1 or num_children == 0):
+                if curr.parent and num_children == 0 and not curr.is_jdbc():  # one table having the prefix
+                    curr = curr.parent
+                assert curr.location is not None
+                external_location = ExternalLocation(curr.location, len(list(curr.all_tables())))
+                external_locations.append(external_location)
                 continue
-            if (
-                not location.startswith("dbfs")
-                and (self._prefix_size[0] < location.find(":/") < self._prefix_size[1])
-                and not location.startswith("jdbc")
-            ):
-                self._dbfs_locations(external_locations, location, min_slash)
-            if location.startswith("jdbc"):
-                self._add_jdbc_location(external_locations, location, table)
-        return external_locations
+            queue.extend(curr.children.values())
+        return sorted(external_locations, key=lambda _: _.location)
+
+    def _resolve_location(self, table: Table) -> Table:
```

Comment:
I think it is confusing that this `resolve_location` returns a `Table` with the resolved location on it. Where `resolve_mount` returns a `str` representing the resolved `mount`

Reply:
Should this become part of the `Table` instead?